### PR TITLE
Data Access State Unit Tests

### PR DESCRIPTION
**Pull request by formalhaunt**
### Description of the Change

Added Unit Tests for the data access classes.

### Why Should This Be In Core?

Testing of the core

### Benefits

More test coverage

### Applicable Issues

https://github.com/constellation-app/constellation/issues/413

### CODE REVIEW

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        );
+    
+        final GraphWriteMethods graph = mock(GraphWriteMethods.class);
+        
+        dataAccessStateIoProvider.readObject(ATTRIBUTE_ID, ELEMENT_ID, jsonNode, graph, null, null, null, null);
+        
+        // Capture the call on the graph setter, pulling out the state
+        final ArgumentCaptor<DataAccessState> captor = ArgumentCaptor.forClass(DataAccessState.class);
+        
+        verify(graph, times(1)).setObjectValue(eq(ATTRIBUTE_ID), eq(ELEMENT_ID), captor.capture());
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java:88]: apparently `verify(graph, times(1))` is the same as `verify(graph)`

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        final ObjectMapper objectMapper = new ObjectMapper();
+        JsonNode root = objectMapper.readTree("null");
+        
+        final GraphWriteMethods graph = mock(GraphWriteMethods.class);
+        
+        dataAccessStateIoProvider.readObject(ATTRIBUTE_ID, ELEMENT_ID, root, graph, null, null, null, null);
+        
+        final ArgumentCaptor<DataAccessState> captor = ArgumentCaptor.forClass(DataAccessState.class);
+        
+        verify(graph, times(1)).setObjectValue(eq(ATTRIBUTE_ID), eq(ELEMENT_ID), captor.capture());
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java:121]: same as above

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        
+        final List<Map<String, String>> state = captor.getValue().getState();
+        
+        assertEquals(state.size(), 0);
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java:125]: `assertTrue(state.isEmpty())`?

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java:143]: similar as above

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        when(graph.isDefaultValue(ATTRIBUTE_ID, ELEMENT_ID)).thenReturn(false);
+        when(graph.getObjectValue(ATTRIBUTE_ID, ELEMENT_ID)).thenReturn(state);
+        
+        final JsonFactory factory = new JsonFactory();
+        final ByteArrayOutputStream output = new ByteArrayOutputStream();
+        JsonGenerator jsonGenerator = factory.createGenerator(output);
+        
+        // The code is written with the assumption that it is called within a document
+        // that has already started being written. Without starting the object in the test
+        // the code would through invalid json exceptions.
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java:171]: throw*

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        when(graph.isDefaultValue(ATTRIBUTE_ID, ELEMENT_ID)).thenReturn(false);
+        when(graph.getObjectValue(ATTRIBUTE_ID, ELEMENT_ID)).thenReturn(null);
+        
+        final JsonFactory factory = new JsonFactory();
+        final ByteArrayOutputStream output = new ByteArrayOutputStream();
+        JsonGenerator jsonGenerator = factory.createGenerator(output);
+        
+        // The code is written with the assumption that it is called within a document
+        // that has already started being written. Without starting the object in the test
+        // the code would through invalid json exceptions.
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java:206]: throw*

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateNGTest.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        dataAccessState = new DataAccessState();
+    }
+
+    @AfterMethod
+    public void tearDownMethod() throws Exception {
+    }
+    
+    @Test
+    public void testConstructorEmptyList() {
+        assertEquals(dataAccessState.getState(), Collections.EMPTY_LIST);
```
Comment (antares1470) [inline CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateNGTest.java:59]: I think `Collections.emptyList()` is better as `Collections.EMPTY_LIST` returns a raw list

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java (hunk 1 above)
Comment (formalhaunt) [reply CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/DataAccessStateIoProviderNGTest.java:89]: I tend to just be a little verbose in my coding sometimes. Not fussed though. I've removed it to match the shortcut.

**Comment by formalhaunt** [comment]
Fixed them all up.

**Review by antares1470 — APPROVED** [review]

**Review by arcturus2 — APPROVED** [review]
Thank you again @formalhaunt for this. I defineatly learnt a lot more of what Mockito can do from this PR.

Note that graph attribute id's can change at any time when a graph is loaded/saved but given you are mocking the `GraphWriteMethods` it makes total sense to have static values for the attributes you used to test `DataAccessStateIoProvider`

**Pull request merged**